### PR TITLE
#223 - Use URL Query Strings for Login Redirect

### LINKED ISSUE #223: Frontend - Better Login Redirect
### Description

From James:

> when I click on a finishline link in Slack on mobile, it will open it in the in-app browser, which never has me logged in already (nor does it ever have NU SSO logged in), so im guaranteed to get bumped to the login page (then google login, then NU SSO). so I always prefer to open it in standalone mobile browser instead, by clicking the button in the bottom right of the in-app browser. the problem here is that the page may load to the login page before i can click that button, meaning that the page being opened into the standalone browser will be the login page, not the page I wanna go to.
theoretically, I think if we made redirect after login just store the soon-to-be URL in a query parameter it would fix this

### Acceptance Criteria

Before being redirected to the login page, the browser should store the intended page to go to in the url like `https://finishlinebyner.com/login?page=projects&value=1.1.0`

After the login is done, it should redirect the user to `https://finishlinebyner.com/projects/1.1.0` by using the query parameters. (I think right now it just goes to what was most recently in your history)

### Proposed Solution

You'll need to poke around with what causes the redirect to the login page in the first place, and in there store the intended page as query parameters right before it redirects.

To deal with query parameters, you can look at how any of the completed tickets have done it in this epic: https://github.com/Northeastern-Electric-Racing/FinishLine/issues/30

### Mocks

_No response_

### PR DESCRIPTION
**Pull request by jonathan-chen10**
## Changes

Modify redirect to login page to use URL query strings ("query args"). Loading `https://finishlinebyner.com/a/b/c?sleep=false&abc=123` without logging in will redirect to `https://finishlinebyner.com/login?page=a&value1=b&value2=c&sleep=false&abc=123`, which will send back to the original page after logging in.

This replaces the `postLoginRedirect` prop for the `Login` component, which is removed.

Hard-coded to ignore a single `/` at the end of a path, so `https://finishlinebyner.com/a/b/c/?sleep=false&abc=123` will redirect to the same login query strings.

## Notes

Because we're using `value1`, `value2`, etc... we can't use these query args anywhere else because then login page might get confused with multiple of the same query and send you to the wrong page after logging in. These are pretty common query string names, so if we use query args a lot we might want to rename these.

## Test Cases

No tests written. Try it manually with the above example.

## To Do

_Any remaining things that need to get done_

- Maybe we want to use [`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) interface to generate the login query strings. May or may not make the code cleaner.

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #223

### CODE REVIEW

File: src/frontend/src/app/AppPublic.tsx [hunk 1]
```diff
@@ -35,10 +35,20 @@ const AppPublic: React.FC = () => {
     }
 
     // otherwise, the user needs to login manually
+    // prepare query args to store path after login
+    const pathParts: string[] = history.location.pathname.split('/').slice(1);
+    // if the URL ended in an extra '/' we don't want to clutter the query args
+    if (pathParts[pathParts.length - 1] === '') pathParts.pop();
```
Comment (anthonybernardi) [inline src/frontend/src/app/AppPublic.tsx:41]: shouldn't this be === '/' ? or am i misreading

File: src/frontend/src/app/AppPublic.tsx [hunk 2]
```diff
@@ -35,10 +35,20 @@ const AppPublic: React.FC = () => {
...
+    const asQueryArg = (pathPart: string, idx: number): string => {
```
Comment (anthonybernardi) [inline src/frontend/src/app/AppPublic.tsx:42]: pretty sure you could just use `reduce` for this

**Review by anthonybernardi — CHANGES_REQUESTED** [review]
this impressive but nearly unreadable

File: src/frontend/src/app/AppPublic.tsx (hunk 1 above)
Comment (jonathan-chen10) [reply src/frontend/src/app/AppPublic.tsx:41]: `'/a/b/c/'.split('/')` gives you `['', 'a', 'b', 'c', '']`

**Review by anthonybernardi — APPROVED** [review]
awesome! this code is always gonna be complicated but this makes a lot more sense

**Pull request merged**